swift/apple/FirezoneKit/Sources/FirezoneKit/Views/UpdateNotification.swift

```diff
@@ -0,0 +1,245 @@
+import UserNotifications
+import Cocoa
+
+struct SemanticVersion: Decodable, Encodable, Comparable {
+  let major: Int
+  let minor: Int
+  let patch: Int
+
+  init(major: Int, minor: Int, patch: Int) {
+      self.major = major
+      self.minor = minor
+      self.patch = patch
+  }
+
+  // This doesn't conform to the full semver spec but it's enough for our use-case
+  private static func parse(versionString: String) -> (major: Int, minor: Int, patch: Int)? {
+      let components = versionString.split(separator: ".")
+      guard components.count == 3,
+            let major = Int(components[0]),
+            let minor = Int(components[1]),
+            let patch = Int(components[2]) else {
+          return nil
+      }
+      return (major, minor, patch)
+  }
+
+  init(from decoder: Decoder) throws {
+      let container = try decoder.singleValueContainer()
+      let versionString = try container.decode(String.self)
+
+      guard let parsed = SemanticVersion.parse(versionString: versionString) else {
+          throw DecodingError.dataCorruptedError(in: container,
+                                                 debugDescription: "Invalid SemVer string format")
+      }
+
+      self.major = parsed.major
+      self.minor = parsed.minor
+      self.patch = parsed.patch
+  }
+
+  static func from(string: String) -> SemanticVersion? {
+      guard let parsed = parse(versionString: string) else {
+          return nil
+      }
+      return SemanticVersion(major: parsed.major, minor: parsed.minor, patch: parsed.patch)
+  }
+
+  func encode(to encoder: Encoder) throws {
+      var container = encoder.singleValueContainer()
+      let versionString = "\(major).\(minor).\(patch)"
+      try container.encode(versionString)
+  }
+
+  static func < (lhs: SemanticVersion, rhs: SemanticVersion) -> Bool {
+      if lhs.major != rhs.major {
+          return lhs.major < rhs.major
+      }
+
+      if lhs.minor != rhs.minor {
+          return lhs.minor < rhs.minor
+      }
+
+      return lhs.patch < rhs.patch
+  }
+
+  static func == (lhs: SemanticVersion, rhs: SemanticVersion) -> Bool {
+      return lhs.major == rhs.major &&
+             lhs.minor == rhs.minor &&
+             lhs.patch == rhs.patch
+  }
+}
+
+private struct VersionInfo: Decodable {
+  let apple: SemanticVersion
+}
+
+private class UpdateNotifier: NSObject, UNUserNotificationCenterDelegate {
+  @Published private(set) var decision: UNAuthorizationStatus
+  private var lastNotifiedVersion: SemanticVersion?
+  static let lastDismissedVersionKey = "lastDismissedVersion"
+  static let notificationIdentifier = "UPDATE_CATEGORY"
+  static let appStoreLink = URL(string: "https://apps.apple.com/us/app/firezone/id6443661826")!
+
+  override public init() {
+    self.decision = .notDetermined
+    super.init()
+
+    let notificationCenter = UNUserNotificationCenter.current()
+
+    // Define a custom action
+    let dismissAction = UNNotificationAction(identifier: "DISMISS_ACTION",
```

Comment:
Can we inline this variable?

Reply:
inline the whole `UNNotificationAction`? I think it's a bit verbose. I do think the string `DISMISS_ACTION` should be extracted, as it's used again to comapre it below.